lumigator/backend/backend/tracking/mlflow.py

```diff
@@ -420,17 +319,280 @@ def list_jobs(self, workflow_id: str):
         )
         return all_jobs
 
+    def _compile_metrics(self, job_ids: list) -> dict:
+        """Take the individual metrics from each run and compile them into a single dict
+        for now, assert that each run has no overlapping metrics
+        """
+        metrics = {}
+        for job_id in job_ids:
+            run = self._client.get_run(job_id)
+            for metric in run.data.metrics:
+                assert metric not in metrics
+                metrics[metric] = run.data.metrics[metric]
+
+        return metrics
+
+    def _compile_parameters(self, job_ids: list) -> dict:
+        """Take the individual parameters from each run and compile them into a single dict
+        for now, assert that each run has no overlapping parameters
+        """
+        parameters = {}
+        for job_id in job_ids:
+            run = self._client.get_run(job_id)
+            for parameter in run.data.params:
+                # if the parameter shows up in multiple runs, prepend the run_name to the key
+                # TODO: this is a hacky way to handle this,
+                #  we should come up with a better solution but at least it keeps the info
+                if parameter in parameters:
+                    parameters[f"{run.data.tags['mlflow.runName']}_{parameter}"] = run.data.params[parameter]
+                parameters[parameter] = run.data.params[parameter]
+        return parameters
+
+    def _find_workflows(self, experiment_id: str) -> list:
+        """Find all the workflows associated with an experiment."""
+        all_runs = self._client.search_runs(experiment_ids=[experiment_id])
+        # now organize the runs into workflows so that a nested run goes under the parent run
+        workflow_ids = []
+        for run in all_runs:
+            parent_id = run.data.tags.get(MLFLOW_PARENT_RUN_ID)
+            # if it has a parent id then it's a run, if not then it's a workflow
+            if parent_id is None:
+                workflow_ids.append(run.info.run_id)
+        return workflow_ids
+
+    def _get_ray_job_logs(self, ray_job_id: str):
```

Comment:
Why do we need the tracking implementation to know about Ray?

Reply:
Yep, `_get_ray_job_logs` was moved from higher up in the original code to here.